chainntnfs/interface_test.go

```diff
@@ -548,6 +548,101 @@ func testBlockEpochNotification(miner *rpctest.Harness,
 	}
 }
 
+// testHistoricalBlockEpochNotification tests the case of multiple registered
+// client receiving historical block epoch  notifications due to their best known
+// block being out of date.
+func testHistoricalBlockEpochNotification(miner *rpctest.Harness,
+	notifier chainntnfs.TestChainNotifier, t *testing.T) {
+
+	const numBlocks = 10
+	const numClients = 5
+	var wg sync.WaitGroup
+	var extraBlockTest sync.WaitGroup
+
+	outdatedHash, outdatedHeight, err := miner.Node.GetBestBlock()
+	if err != nil {
+		t.Fatalf("unable to retrieve current height: %v", err)
+	}
+
+	// Generate numBlocks blocks which clients should receive historical
+	// notifications for.
+	if _, err := miner.Node.Generate(numBlocks); err != nil {
+		t.Fatalf("unable to generate blocks: %v", err)
+	}
+
+	// Ensure the neutrino notifier's best height is caught up to the chain.
+	// Otherwise, the clients will not receive all of the historical block
+	// notifications.
+	switch notifier := notifier.(type) {
+	case *neutrinonotify.NeutrinoNotifier:
+		targetHeight := uint32(outdatedHeight + numBlocks)
+		for totalMs := 0; notifier.GetBestHeight() != targetHeight &&
+			totalMs < 5000; totalMs += 100 {
+
+			time.Sleep(100 * time.Millisecond)
+		}
+		if notifier.GetBestHeight() != targetHeight {
+			t.Fatalf("neutrino best height did not catch up to chain")
+		}
+	}
+
+	// Create numClients clients whose best known block is 10 blocks behind
+	// the tip of the chain. We expect each client to receive numBlocks
+	// notifications, 1 for each block  they're behind. So we'll use a
+	// WaitGroup to synchronize the test.
```

Comment:
I'm wondering if this could be simplified a little bit...

```go
foreach client {
   for numBlocks { 
      select {
       case block <- client.Epochs:
            check...
       case <-time.After(timeout):
             fail()
       }
   }

   wg.Add(1)
   go func() {
       defer wg.Done()
       check no unexpected notification timeout or error after 5 sec
   }()
}

wg.Wait()
```

Reply:
I think this would cause a deadlock, since the channels are buffered to 20 (i think). The client loop would need to be the inner so they all pull blocks together

Reply:
Although it looks like the test is only 10 blocks, so could still work

Reply:
Oh, so we can actually deadlock in the notifier itself if the client is not reading its notifications? Is that something we should try to handle?

Reply:
fixed, I made the client loop the inner

Reply:
@valentinewallace sweet!

@halseth i think it's possible, though we've never encountered it. Could change to using a concurrent queue, though then it's also possible to have unbounded memory. Probably out of scope for this PR though